### PR TITLE
Handle the case when Float64(p) + Float64(q) != 1 in gamma_inc_inv

### LINKED ISSUE #373: gamma_inc_inv fails for low precision floats
See the error below. The problem is that we currently promote the arguments to `Float64` and then promote back but generally `Float64(p) + Float64(p) != 1` when `p + q == 1` if `p` and `q` have lower precision than `Float64` and that will trigger an exception.
```julia
julia> gamma_inc_inv(1.0f0, 0.9f0, 0.1f0)
ERROR: ArgumentError: p + q must equal one but is 0.9999999776482582
Stacktrace:
 [1] _gamma_inc_inv(a::Float64, p::Float64, q::Float64)
   @ SpecialFunctions ~/.julia/dev/SpecialFunctions/src/gamma_inc.jl:927
 [2] _gamma_inc_inv
   @ ~/.julia/dev/SpecialFunctions/src/gamma_inc.jl:1017 [inlined]
 [3] gamma_inc_inv(a::Float32, p::Float32, q::Float32)
   @ SpecialFunctions ~/.julia/dev/SpecialFunctions/src/gamma_inc.jl:922
 [4] top-level scope
   @ REPL[155]:1
```
Maybe the solution is to change the `Float16` and `Float32` methods to compute `1 - Float64(min(p, q))` and verify that this number is identical to the input in the original precision before calling the `Float64` method.

### PR DESCRIPTION
**Pull request by andreasnoack**
for lower precision p and q.

Fixes #373

### CODE REVIEW

**Comment by stevengj** [comment]
Wouldn't it be better to simply have a one-argument `gamma_inc_inv` method?  Why are we passing two arguments when one of them is computable from the other?

File: src/gamma_inc.jl [hunk 1]
```diff
@@ -1014,8 +1014,22 @@ function _gamma_inc_inv(a::Float64, p::Float64, q::Float64)
...
+            throw(ArgumentError("p + q must equal one but was $(p + q)"))
+        end
+        Float64(p), _q
+    else
+        _p = 1 - Float64(q)
+        if p != T(_p)
+            throw(ArgumentError("p + q must equal one but was $(p + q)"))
+        end
+        _p, Float64(q)
+    end
```
Comment (stevengj) [inline src/gamma_inc.jl:1030]: Why not just
```jl
p + q == one(T) || throw(ArgumentError("p + q must equal one but was $(p + q)"))
p64, q64 = p < q ? (Float64(q), 1 - Float64(p)) : (1 - Float64(q), Float64(q))
```

**Review by devmotion — COMMENTED** [review]
Isn't this still problematic? Couldn't floating point issues result in `(1 - p) + p != 1`?

Maybe the following could be an alternative:
- Check `p + q == 1` with the provided inputs without promoting
- Check special cases `iszero(p)` and `iszero(q)` (still without promotion, currently in the Float64 implementation)
- Call an internal method with only `Float64(min_p_q)` (and if really needed a second argument `1 - Float64(min_p_q)` but would be nice to avoid this to not introduce any floating point issues)

File: src/gamma_inc.jl (hunk 1 above)
Comment (andreasnoack) [reply src/gamma_inc.jl:1030]: Yeah. That avoids some replication. I'll change it to something along these lines. However, I find it easier to read with the branches on separate lines and it also makes it easier to verify coverage.

**Comment by andreasnoack** [comment]
> Isn't this still problematic? Couldn't floating point issues result in `(1 - p) + p != 1`?

I actually don't know the answer to that question by I'm pretty sure that's quite rare. I'd like to proceed with the current approach for now. It's a pretty small change and the current versions currently only work in very special cases for Float16s and Float32s.

**Comment by devmotion** [comment]
Yes, seems reasonable to keep the number of changes small in this PR. I think one could refactor the methods a bit but probably best to do this in a separate PR.

**Comment by andreasnoack** [comment]
Bump. This one is needed over at https://github.com/JuliaStats/StatsFuns.jl/pull/113 so it would be great to get merged.

**Comment by stevengj** [comment]
> Isn't this still problematic? Couldn't floating point issues result in `(1 - p) + p != 1`?

**Review by stevengj — APPROVED** [review]

**Comment by simonbyrne** [comment]
> I don't think we can have `(1 - p) + p != 1` for `0 ≤ p ≤ 1`, though I don't have a proof. In any case, this PR seems like a strict improvement over the previous behavior.

By exhaustive testing, it's true at least for Float32:
```julia
function check()
  for u = reinterpret(UInt32, 0.0f0):reinterpret(UInt32, 1.0f0)
    f = reinterpret(Float32,u)
    (1-f) + f == 1 || @show f
  end
end
```

Roughly, I think the proof would rely on the fact that the rounding error incurred by `cf = 1-f` should have the opposite sign to the rounding error incurred by `cf + f`, which should give you a reasonable bound on the total error.

**Pull request merged**